Added support for AMD's Zen3 architecture

dzambare opened the pull request:
Added support for Zen3 Configuration.

 -- Added new zen3 configuration and auto detection of zen3 architecture 
 -- Added configuration family amd64 for all zen architectures
 -- Moved older AMD architecture in amd64_legacy family
 -- Updated zen2 makefiles to pick znver2 for clang (if supported by the detected clang version)

----------------------------------------------------------------------------------------------------------------------
AMD BLIS Upstream:

This PR includes following commits for AMD BLIS version 3.0.1

pick 9c7814da Added support for zen3 configuration
squash 536edc40 Added support for zen3 configuration
squash 23a2073c Added support for zen3 configuration
squash 449ee370 Added support for zen3 configuration
pick 25d23cdd Zen3 support, disabled IR, JR loop parallelization
squash 9d7978ee Zen3 support, disabled IR, JR loop parallelization
pick f8ab9f63 Enabled znver3 flag for zen3 architecture
squash 38a8008c Enabled znver3 flag for zen3 architecture
pick b1144a85 Added -fomit-frame-pointer option to CKOPTFLAGS.
pick ce99b1ec Added dynamic block size selection logic for DGEMM.
squash f9d06c74 Added dynamic block size selection logic for DGEMM.
pick 14e21603 Update amd64 bundle configuration
squash c2f63fcc Update amd64 bundle configuration

dzambare commented:
> @fgvanzee , generic_legacy it is failing because there is issue with amd64_legacy configuration, interestingly amd64_legacy builds fine on its own but fails via x86_64.
> 

Seems like it some name conflict, if  I rename **amd64_legacy** to **amd_legacy** it works but **amd64legacy** doesn't. Do we ignore characters after certain length in the config name?

fgvanzee commented:
I've found the source of the `amd64_legacy` / `generic_legacy` bug. It is indeed the result of a poorly-formed `sed` substitution.  I'll try to commit a fix soon.

fgvanzee commented:
@devinamatthews The fix *should* have worked, though. Instead, the `x86_64` build is still failing, but in a way that I don't understand. Can you [take a look](https://app.travis-ci.com/github/flame/blis/jobs/543296534)?

fgvanzee commented:
This [link](https://www.intel.com/content/dam/develop/external/us/en/documents/downloads/sde-external-8.63.0-2021-01-18-lin.tar.bz2)?

fgvanzee commented:
> Uh, isn't the fix the check for a high enough version in the make defs? No need to change the Travis CI setup.

For the record, the `make_defs.mk` is already trying to perform version logic. I guess there is something wrong with it?
```sh
CKOPTFLAGS     := $(COPTFLAGS)
ifeq ($(CC_VENDOR),gcc)
GCC_VERSION := $(strip $(shell gcc -dumpversion | cut -d. -f1))
#gcc or clang version must be atleast 4.0
# gcc 9.0 or later:
ifeq ($(shell test $(GCC_VERSION) -ge 9; echo $$?),0)
CKVECFLAGS     += -march=znver2
else
# If gcc is older than 9.1.0 but at least 6.1.0, then we can use -march=znver1
# as the fallback option.
CRVECFLAGS += -march=znver1 -mno-avx256-split-unaligned-store
CKVECFLAGS += -march=znver1 -mno-avx256-split-unaligned-store
endif
else
ifeq ($(CC_VENDOR),clang)
```

dzambare commented:
> The problem looks like it is unconditionally using `gcc` instead of the actual chosen compiler. It would be much better to have configure deal with getting the version and then make this available to the Makefiles.

Sorry for delay in reply, and thanks for investigating the issue. 

Zen3 make_def.mk does use correct flags based on compiler version, but as pointed out by @devinamatthews the issue is using the compile in path ('gcc') rather than using one used during configure ($(CC)). I have it fixed in later AMD releases but did miss using GCC_OT_* variables. 

For this PR I will update it to use $(CC) in config/zen[2,3]/make_defs.mk as we have the logic for both gcc and clang in make-def.mk. 
Or do you want me to update configure to set CLANG_OT_* ? and use them here?

dzambare commented:
@fgvanzee, @devinamatthews, let me know if you have any other inputs, change request. 
If we want to add CLANG_OT_? I feel I should do it in separate PR anyways for clarity and separation, what do you say?

fgvanzee commented:
@dzambare I'd like to make changes to the PR. Let me know when you've made whatever updates you had planned.

A few disparate but related comments:
- I'm not sure if I want to defer `CLANG_OT_*` to a future commit. (It depends on whether we keep the AOCC version detection.)
- Vanilla BLIS currently does not recognize AOCC as a separate compiler.
- If we do support AOCC in vanilla BLIS, I want the compiler version detection to happen in `configure`.

fgvanzee commented on config/zen3/make_defs.mk:
```diff
@@ -0,0 +1,137 @@
+#gcc or clang version must be atleast 4.0
+# gcc 9.0 or later:
+ifeq ($(shell test $(GCC_VERSION) -ge 9; echo $$?),0)
+CKVECFLAGS     += -march=znver2
```
Good eye, @bartoldeman. I had noticed this, too, and have fixed it in my (as yet unpushed) pass of edits.

fgvanzee commented:
One thing that will help me get started on the `CLANG_OT_*` variables, @dzambare, is if you could enumerate all of the important version thresholds needed by the `zen2` and `zen3` subconfigs. (If you want to include `zen` too, you can.)

Another question: can we distinguish the important AOCC versions through its clang versions (since AOCC is based on clang), or do you think we will need a separate set of `AAOC_OT_*` variables to properly distinguish between important groupings of versions of AOCC?

devinamatthews commented:
Why not just check the compiler for support for various flags in `configure`? (or heck, you could do it in the Makefile if you wanted)

fgvanzee commented:
> Why not just check the compiler for support for various flags in `configure`? (or heck, you could do it in the Makefile if you wanted)

I'm not sure how this is incompatible with what I've proposed.

devinamatthews commented:
For example:

```bash
has_znver3=no
if echo | $CC -march=znver3 -x c -c -o /dev/null -; then
    has_znver3=yes
fi
```

Then sub $HAS_ZNVER3@ or something in config.mk.

devinamatthews commented:
And of course you want to add `2> /dev/null` to catch the potential error message.

devinamatthews commented:
The conflict is that then you would not need any version flags at all.

fgvanzee commented:
> The conflict is that then you would not need any version flags at all.

You wouldn't need any versions variables, true, but this solution still creates new `make` variables that need to be communicated to the makefile fragment. You're just changing what information is communicated.

I like your solution in principle since it narrowly identifies the key piece of information, which is: "Does the compiler support `-march=znver3`? But I would want to switch to this system of tracking options (rather than versions) uniformly for all subconfigs. It's also not clear how that would work in the context of the older versions of Intel flags (`-march=core-avx2`). Do newer compilers still support those older flags? Or maybe we don't care if it is supposed to work either way.

I could also imagine someone arguing that versions are still a bit more useful because in theory you could have a version of a compiler that *thinks* it supports a compiler flag, but that support is actually broken in practice.

dzambare commented:
> One thing that will help me get started on the `CLANG_OT_*` variables, @dzambare, is if you could enumerate all of the important version thresholds needed by the `zen2` and `zen3` subconfigs. (If you want to include `zen` too, you can.)
> 
> Another question: can we distinguish the important AOCC versions through its clang versions (since AOCC is based on clang), or do you think we will need a separate set of `AAOC_OT_*` variables to properly distinguish between important groupings of versions of AOCC?

For AOCC (AMD clang) we have following thresholds:

AOCC > 3.0.0  supports znver3.
AOCC > 2.0.0  supports znver2.
AOCC < 2.0.0  supports znver1.

For clang I have limited information 

clang > 9.0.0 supports znver2
clang < 9.0.0 supports znver1

dzambare commented:
> > The conflict is that then you would not need any version flags at all.
> 
> You wouldn't need any versions variables, true, but this solution still creates new `make` variables that need to be communicated to the makefile fragment. You're just changing what information is communicated.
> 
> I like your solution in principle since it narrowly identifies the key piece of information, which is: "Does the compiler support `-march=znver3`? But I would want to switch to this system of tracking options (rather than versions) uniformly for all subconfigs. It's also not clear how that would work in the context of the older versions of Intel flags (`-march=core-avx2`). Do newer compilers still support those older flags? Or maybe we don't care if it is supposed to work either way.
> 
> I could also imagine someone arguing that versions are still a bit more useful because in theory you could have a version of a compiler that _thinks_ it supports a compiler flag, but that support is actually broken in practice.

Me too like this idea, specifically when we are dealing with vanilla clang and AMD clang it will avoid many redundant checks/variables.

devinamatthews commented:
Clang 12+ supports the `-znver3` flag.

fgvanzee commented:
@dzambare Please take a look at 7872c3a -- in particular, the compiler detection logic in `configure` (starting around line 1438), and the `make_defs.mk` files for the `zen2` and `zen3` subconfigs. If you have time, it would also be good to perform some quick tests in your development environment for `zen3` with both `clang` and AOCC to make sure that `configure` performs as intended.

I also wanted to add CPUID files for Zen2 and Zen3 in `travis/cpuid`, but I'm not confident that I would create valid files. So, for now, we are only able to test Zen1 via the SDE in the Travis CI builds. I might be able to return to this later someday, but if you wanted to take a stab at this you are welcome to.

fgvanzee commented:
@dzambare I made a minor change of note in b641cf7: I added `-mllvm -disable-licm-vrp` back to the list of compiler options for Zen1 with AOCC in the `zen2` subconfig. These options were previously included via `amd_config.mk`, but I forgot to include them in my rewrite.

I also rewrote the `zen/make_defs.mk` file to match that of the `zen2` and `zen3` fragments.

dzambare commented:
[AMD Official Use Only]

Hi Field,

Thanks, I will review and test with AOCC and get back with observations and feedback.

Regards,
Dipal
________________________________
From: Field G. Van Zee ***@***.***>
Sent: Thursday, November 11, 2021 1:15:05 AM
To: flame/blis ***@***.***>
Cc: Zambare, Dipal Madhukar ***@***.***>; Mention ***@***.***>
Subject: Re: [flame/blis] Added support for AMD's Zen3 architecture (#561)

[CAUTION: External Email]

@dzambare<https://nam11.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fdzambare&data=04%7C01%7Cdipalmadhukar.zambare%40amd.com%7C1992e74c00a3410d21dc08d9a482986c%7C3dd8961fe4884e608e11a82d994e183d%7C0%7C0%7C637721703077672239%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=FNIfMO8F8AnrYhMZqGy6ePr04LWTAwN%2FXUqFXyEMMvY%3D&reserved=0> I made a minor change of note in b641cf7<https://nam11.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fflame%2Fblis%2Fcommit%2Fb641cf7c0972ae1d4eeb9081bfe79464a0e382fe&data=04%7C01%7Cdipalmadhukar.zambare%40amd.com%7C1992e74c00a3410d21dc08d9a482986c%7C3dd8961fe4884e608e11a82d994e183d%7C0%7C0%7C637721703077682233%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=EMGppdKvlnt03s%2BP0oHb9chWxEP2YhXzYzmYzxYjQW8%3D&reserved=0>: I added -mllvm -disable-licm-vrp back to the list of compiler options for Zen1 with AOCC in the zen2 subconfig. These options were previously included via amd_config.mk, but I forgot to include them in my rewrite.

I also rewrote the zen/make_defs.mk file to match that of the zen2 and zen3 fragments.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://nam11.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fflame%2Fblis%2Fpull%2F561%23issuecomment-965684186&data=04%7C01%7Cdipalmadhukar.zambare%40amd.com%7C1992e74c00a3410d21dc08d9a482986c%7C3dd8961fe4884e608e11a82d994e183d%7C0%7C0%7C637721703077682233%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=fm8KYjyFxw8L5BH4xqZ%2BB5aufD5LBAUioR3zZL4hpFU%3D&reserved=0>, or unsubscribe<https://nam11.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FARAPQ7C7TZMNU2V2RZG7DEDULLDUDANCNFSM5F7ELVMQ&data=04%7C01%7Cdipalmadhukar.zambare%40amd.com%7C1992e74c00a3410d21dc08d9a482986c%7C3dd8961fe4884e608e11a82d994e183d%7C0%7C0%7C637721703077692230%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=446ozBnBlos2t32ZtUueA2iQq9r2RCeXSOvgPr8dtq0%3D&reserved=0>.
Triage notifications on the go with GitHub Mobile for iOS<https://nam11.safelinks.protection.outlook.com/?url=https%3A%2F%2Fapps.apple.com%2Fapp%2Fapple-store%2Fid1477376905%3Fct%3Dnotification-email%26mt%3D8%26pt%3D524675&data=04%7C01%7Cdipalmadhukar.zambare%40amd.com%7C1992e74c00a3410d21dc08d9a482986c%7C3dd8961fe4884e608e11a82d994e183d%7C0%7C0%7C637721703077702225%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=CATiLiYl630OHrhwwz0Kay4egFrvSC4RZMORTm0c0ek%3D&reserved=0> or Android<https://nam11.safelinks.protection.outlook.com/?url=https%3A%2F%2Fplay.google.com%2Fstore%2Fapps%2Fdetails%3Fid%3Dcom.github.android%26referrer%3Dutm_campaign%253Dnotification-email%2526utm_medium%253Demail%2526utm_source%253Dgithub&data=04%7C01%7Cdipalmadhukar.zambare%40amd.com%7C1992e74c00a3410d21dc08d9a482986c%7C3dd8961fe4884e608e11a82d994e183d%7C0%7C0%7C637721703077702225%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=l8JVOnbWzrh%2Fb0%2FEtkkgxgaFxLrzHwmteyrYa%2FfWzC4%3D&reserved=0>.

fgvanzee commented:
@dzambare Thanks Dipal. Devin provided me guidance on the CPUID `.def` files, so no need to investigate that yet. I might be able to take care of it myself, but I may wait and do it outside of this PR.

fgvanzee commented:
@dzambare Looks like the SDE testing for `zen2` and `zen3` subconfigurations is working.

dzambare commented:
Hi @fgvanzee, I have tested the new makefiles for gcc 9, 10 , 11, aocc 2.0, 3.0 and 3.1 and clang 11. Everything is working as expected. 

Really appreciate your help in updating makefiles and SDE.

fgvanzee commented:
You're welcome, @dzambare. Thank you for your help as well! Glad the new configuration logic works as intended.

fgvanzee commented:
@bartoldeman This (rather significant) PR merge is done. :slightly_smiling_face:
